### PR TITLE
[ML] Complete the Data Frame task on stop

### PR DESCRIPTION
**Pull request by davidkyle**
A call to `stop()` will cause the indexer to stop shortly after when the next search returns. The persistent task can only be completed once the indexer is stopped otherwise we have a zombie indexer spinning away but not accessible to the task manager. If the `wait_for_completion` flag is true the stop requests waits for the persistent task to complete.

I added a `onStop` event which is fired when the indexer stops. Note this is different to `onFinish` which is triggered when the indexer completes the current phase. 

Delete is changed to simply check if the task is active and if not delete the data frame transform config.

### CODE REVIEW

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportDeleteDataFrameTransformAction.java [hunk 1]
```diff
@@ -5,93 +5,72 @@
...
-            }, listener::onFailure));
+    protected void masterOperation(Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener) throws Exception {
+        PersistentTasksCustomMetaData pTasksMeta = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
+        if (pTasksMeta != null && pTasksMeta.getTask(request.getId()) != null) {
+            listener.onFailure(new ElasticsearchStatusException("Cannot delete data frame [" + request.getId() +
+                    "] as the task is running. Stop the task first", RestStatus.CONFLICT));
         } else {
-            listener.onFailure(new IllegalStateException("Could not delete transform [" + request.getId() + "] because "
-                    + "indexer state is [" + state + "].  Transform must be [" + IndexerState.STOPPED + "] before deletion."));
+            // Task is not runing, delete the configuration document
```
Comment (benwtrent) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportDeleteDataFrameTransformAction.java]:
```suggestion
            // Task is not running, delete the configuration document
```

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportDeleteDataFrameTransformAction.java [hunk 2]
```diff
@@ -5,93 +5,72 @@
...
-            // on the master, which makes testing difficult.
-            if (nodes.getMasterNode() == null) {
-                listener.onFailure(new MasterNotDiscoveredException("no known master nodes"));
-            } else {
-                transportService.sendRequest(nodes.getMasterNode(), actionName, request,
-                        new ActionListenerResponseHandler<>(listener, Response::new));
-            }
-        }
+    protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+        return null;
```
Comment (benwtrent) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportDeleteDataFrameTransformAction.java]: Wouldn't it be a good idea to make sure that the `metadata` read information is not dirty before executing this task?
```
return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
```

**Review by benwtrent — COMMENTED** [review]
Since we are removing the persistent task from cluster state when `stop` is called, there is no way to start it again from where the cursor is currently located. 

Are there plans to check the `IndexerState` when `onSaveState` is called and determining if a indexed snapshot of the current cursor position and checkpoint should be made? 

I think something like:
```
if (indexerState.equals(IndexerState.DELETING)) { ...take snapshot...}
```
would be a good idea. That way when `start` is called again, the previous cursor position can be loaded up from the index.

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportDeleteDataFrameTransformAction.java (hunk 1 above)
Comment (davidkyle) [reply x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportDeleteDataFrameTransformAction.java]: 👍

**Comment by davidkyle** [comment]
There was a failure in test clean up after the multi node tests
```ElasticsearchStatusException[Cannot delete data frame [data-frame-transform-crud] as the task is running. Stop the task first]```

I made the stop action delegate to the master node for coordinating responses which should fix the problem

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/DataFrameTransformDocumentationIT.java [hunk 3]
```diff
@@ -76,7 +76,8 @@
     @After
     public void cleanUpTransforms() throws IOException {
         for (String transformId : transformsToClean) {
-            highLevelClient().dataFrame().stopDataFrameTransform(new StopDataFrameTransformRequest(transformId), RequestOptions.DEFAULT);
+            highLevelClient().dataFrame().stopDataFrameTransform(
+                    new StopDataFrameTransformRequest(transformId, Boolean.TRUE, TimeValue.timeValueSeconds(20)), RequestOptions.DEFAULT);
```
Comment (benwtrent) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/DataFrameTransformDocumentationIT.java]: `Boolean.True` should probably just be `true`

**Review by benwtrent — APPROVED** [review]
We should consider taking snapshots of the current cursor position and checkpoint so that transforms can be stopped, and then started again from the same spot.

**Comment by davidkyle** [comment]
run elasticsearch-ci/1

**Comment by davidkyle** [comment]
run elasticsearch-ci/1

**Pull request merged**